Fix inconsistent sound effects on mod buttons

bdach opened the pull request:
Fixes #5916.

Because `HoverClickSounds.OnClick()` does not fire upon right clicking on mod buttons, the sound effects that play on left and right click were inconsistent. Introduce `HoverMouseUpSounds` drawable that allows to play the click sound effect upon mouse up events for an arbitrary set of mouse buttons and use it on mod buttons.

FWIW, this works, but it feels a little hacky, so if this is not the way you'd like the issue fixed, feel free to close this.

bdach committed:
Fix inconsistent sound effects on mod buttons

Because HoverClickSounds.OnClick() does not fire upon right clicking
on mod buttons, the sound effects that play on left and right click
were inconsistent. Introduce HoverMouseUpSounds drawable that allows
to play the click sound effect upon mouse up events for an arbitrary
set of mouse buttons and use it on mod buttons.

iiSaLMaN commented:
Can't this be done by adding a field to `HoverClickSounds` that would allow right clicks? or adding a constructor parameter to `HoverClickSounds` and make it optional `new[] { MouseButton.Left }` which might be better in case we want click sounds for middle mouse button.

bdach commented:
I'm not honestly sure, but I don't think that's possible without any deeper changes. As I mentioned in the issue, `HoverClickSounds.OnClick(ClickEvent)` doesn't even fire when right click is used to toggle mods - there is no inherent filtering by button in there, so it doesn't appear the event is even propagated to that method.

I admit I don't know the input-handling code, but it feels that either the event is filtered/handled somewhere higher in the drawable tree, or the `OnClick()` handler _semantically_ means left click and left click only, in which case I don't think we even want it to get right-click events because _a lot_ of things could start acting funny.

iiSaLMaN commented:
Simply what I mean is to handle that on `OnMouseUp`,  because it's equivalent to `OnMouseClick` but for any other mouse button.

bdach commented:
Oh, that's what you meant - sure, that would also work if that's the preferred approach.

bdach committed:
Handle other button clicks in HoverClickSounds

As suggested in review, remove previously introduced HoverMouseUpSounds
and instead change effect playing logic in HoverClickSounds by moving it
out of OnClick() to OnMouseUp().

Users of the class can either use the existing constructor to play
the effect only on left click or use the newly introduced constructor
with the MouseButton[] parameter to specify which button clicks should
trigger the sound.

iiSaLMaN commented on osu.Game/Graphics/UserInterface/HoverClickSounds.cs:
```diff
@@ -16,16 +18,38 @@ namespace osu.Game.Graphics.UserInterface
+        /// </summary>
+        /// <param name="buttons">Array of button codes which should trigger the click sound.</param>
+        /// <param name="sampleSet">Set of click samples to play.</param>
+        public HoverClickSounds(MouseButton[] buttons, HoverSampleSet sampleSet = HoverSampleSet.Normal)
```
```suggestion
        public HoverClickSounds(MouseButton[] buttons = new[] { MouseButton.Left }, HoverSampleSet sampleSet = HoverSampleSet.Normal)
```

bdach replied:
This won't actually compile, a default value parameter must be a compile-time constant. Best I could do here is default to `null` and then coalesce:
```csharp
public HoverClickSounds(MouseButton[] buttons = null, HoverSampleSet sampleSet = HoverSampleSet.Normal)
    : base(sampleSet)
{
    this.buttons = buttons ?? new[] { MouseButton.Left };
}
```

iiSaLMaN replied:
Better than another constructor.

bdach committed:
Consolidate HoverClickSounds constructors

As suggested in review, merge both HoverClickSounds constructors into
one accepting optional arguments. Due to existing usages the parameter
is added as second and supplied by name in ModButton.

bdach committed:
Fix inaccurate xmldoc

frenzibyte reviewed (dismissed):
LGTM 👍

peppy commented on osu.Game/Graphics/UserInterface/HoverClickSounds.cs:
```diff
@@ -16,16 +18,32 @@ namespace osu.Game.Graphics.UserInterface
         {
-            sampleClick?.Play();
-            return base.OnClick(e);
+            var index = Array.IndexOf(buttons, e.Button);
```
please use `Contains` instead.

bdach committed:
Replace Array.IndexOf() with Contains()

peppy committed:
Merge remote-tracking branch 'upstream/master' into consistent-mod-button-sounds

peppy committed:
Simplify implementation

peppy committed:
Improve xmldoc
